src/buffer/out/Row.cpp

```diff
@@ -1181,3 +1181,36 @@ CharToColumnMapper ROW::_createCharToColumnMapper(ptrdiff_t offset) const noexce
     const auto guessedColumn = gsl::narrow_cast<til::CoordType>(clamp(offset, 0, _columnCount));
     return CharToColumnMapper{ _chars.data(), _charOffsets.data(), lastChar, guessedColumn };
 }
+
+const std::optional<ScrollbarData>& ROW::GetScrollbarData() const noexcept
+{
+    return _promptData;
+}
+void ROW::SetScrollbarData(std::optional<ScrollbarData> data) noexcept
+{
+    _promptData = data;
+}
+
+void ROW::StartPrompt() noexcept
+{
+    if (!_promptData.has_value())
+    {
+        _promptData = ScrollbarData{
+            .category = MarkCategory::Prompt,
+            .color = std::nullopt,
+            .exitCode = std::nullopt,
+        };

```

Comment:
FYI, and I'm writing this just because I think it's interesting, as I mentioned elsewhere, `optional` is really only good at keeping stuff uninitialized. It's not really consistently great at initializing things, however.

As Dustin said, an `optional` is basically
```cpp
template<typename T>
struct optional {
    union { T data }; // union keeps it uninitialized even if it has a constructor
    bool initialized = false;
};
```

The problem with `optional` is that its constructor
```
template < class U = T >
constexpr optional( U&& value );
```
takes the initial value by reference, `std::move`s it into `data` and sets `initialized` to `true`. This _only_ works great if the compiler happens to figure out how to optimize it. This doesn't work consistently in any compiler, but MSVC in particular _also_ happens to be really _**REALLY**_ bad at avoiding unnecessary struct copies.

This means that constructing a `ScrollbarData` like you do looks like this:
```asm
mov     BYTE PTR [rcx], 4            ; .category = 4
mov     BYTE PTR [rcx+8], 0          ; .color.initialized = false
mov     BYTE PTR [rcx+16], 0         ; .exitCode.initialized = false
```

However, constructing a `std::optional<ScrollbarData>` for `_promptData` looks like this:
```asm
; $T1 == the current function stack
mov     BYTE PTR $T1[rsp], 4         ; stack: .category = 4
mov     BYTE PTR $T1[rsp+8], 0       ; stack: .color.initialized = false
movups  xmm0, XMMWORD PTR $T1[rsp]   ; stack -> register
mov     BYTE PTR $T1[rsp+16], 0      ; stack: .exitCode.initialized = false
mov     eax, DWORD PTR $T1[rsp+16]   ; stack -> register
movups  XMMWORD PTR [rcx], xmm0      ; register -> ROW (.category/.color)
mov     DWORD PTR [rcx+16], eax      ; register -> ROW (.exitCode)
mov     BYTE PTR [rcx+20], 1         ; ._promptData.initialized = true
```

In other words, MSVC constructs the `ScrollbarData` on stack, then copies that into register, and then copies it again into `ROW`. Other compilers don't do this exact thing, but they're doing similar dumb stuff under similar circumstances.

However, fret not, because the C++ consortium in their infinite wisdom have solved this problem ~by making improvements to the language like introducing proper union types~ by adding another overload to `std::optional` which can be used here:
```
template< class... Args >
constexpr explicit optional( std::in_place_t, Args&&... args );
```

```cpp
std::optional<ScrollbarData>{
    std::in_place,
    MarkCategory::Prompt,
    std::nullopt,
    std::nullopt,
};
```

Just kidding:
```asm
mov     BYTE PTR $T1[rsp], 4
mov     BYTE PTR $T1[rsp+8], 0
movups  xmm0, XMMWORD PTR $T1[rsp]
mov     BYTE PTR $T1[rsp+16], 0
mov     BYTE PTR $T1[rsp+20], 1
movsd   xmm1, QWORD PTR $T1[rsp+16]
movups  XMMWORD PTR [rcx], xmm0
movsd   QWORD PTR [rcx+16], xmm1
```

This occurs because fixing the construction doesn't fix the existence of its move constructor. That one is still implemented in the STL and thus relies on the compiler optimizing it which no compiler does consistently correct.

The actual proper fix is to use this:
```cpp
if (!_promptData.has_value())
{
    _promptData.emplace(MarkCategory::Prompt);
}
```

which results in this:
```asm
cmp     BYTE PTR [rcx+20], 0         ; _promptData.has_value()
jne     SHORT $LN8@test              ; if (!...) goto ret;
mov     BYTE PTR [rcx], 4            ; .category = 4
mov     BYTE PTR [rcx+8], 0          ; .color.initialized = false
mov     BYTE PTR [rcx+16], 0         ; .exitCode.initialized = false
mov     BYTE PTR [rcx+20], 1         ; ._promptData.initialized = true
$LN8@test:
ret     0
```

While using `emplace` fixes the issue here, my point is mainly that `optional` is not particularly great for performance nor memory usage.